consensus/reactor.go

```diff
@@ -1074,8 +1075,8 @@ func (ps *PeerState) GetRoundState() *cstypes.PeerRoundState {
 func (ps *PeerState) MarshalJSON() ([]byte, error) {
 	ps.mtx.Lock()
 	defer ps.mtx.Unlock()
-
-	return cmtjson.Marshal(ps)
+	type aPeerState PeerState
```

Comment:
I believe you are following the strategy similar to the one described [here](https://jhall.io/posts/go-json-self-referencing-marshaler/).

If that trick breaks the infinite loop, can we still use `cmtjson`?  (to avoid further unknown/unintended side-effects)

Reply:
yes, like other type is [toAllRoundVotes](onlyhttps://github.com/cometbft/cometbft/blob/main/consensus/types/height_vote_set.go#L260)

Reply:
Cool, I think I have something. The problems I see there:
* Let's not move away from `cmtjson`
* Let's not copy already-initialized mutexes

I've fixed that on a branch (using this PR as starting point).
Please cherry-pick `8aa5ee11a..eac18b3a9` if you agree with those changes.

With those changes, the PR would be ready for approval (AFAIAC)

Reply:
Got it thanks, implements MarshalJSON need to be careful, even without lock might be recursive

Reply:
My concern is that, if you copy the `PeerState`, rather than a pointer to it, you are effectively copying its internal Mutex. Documentation of `sync.Mutex` says that is _wrong_ (bear in mind `sync.Mutex` is a struct, not an interface)

Reply:
not sure copy properties one by one is a better way